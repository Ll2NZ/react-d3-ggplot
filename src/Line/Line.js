import React               from "react";
import { Component }       from "react";
import PropTypes           from "prop-types";
import { line }            from "d3-shape";
import { curveCatmullRom } from "d3-shape";
import { select }          from "d3-selection";
import { scaleFinder }     from "../Utilities/";

class Line extends Component {
    static defaultProps = {
        dimensions:
        {
            width: window.innerWidth*0.9,
            height: window.innerHeight*0.9,
            padding: 50
        },
        className: null,
        color: "orange",
        lineWidth: 1
    }

    static propTypes = {
        data: PropTypes.array.isRequired,
        dimensions: PropTypes.shape({
            width: PropTypes.number,
            height: PropTypes.number,
            padding: PropTypes.number
        }),
        scaleTypes: PropTypes.array,
        className: PropTypes.string,
        aes: PropTypes.array,
        color: PropTypes.string,
        lineWidth: PropTypes.number
    }

    getXScale(){
        // get props and aesthetic to render
        const { data, aes } = this.props;

        // get x-values
        const xValues  = data.map(item => item[aes[0]]);
        const scaleObj = new scaleFinder(xValues);

        return scaleObj;
    }

    getXScaleType(){
        const { dimensions, scaleTypes } = this.props;
        const scaleObj                   = this.getXScale();
        let xScale;

        if(scaleTypes[0] === "linear")
            xScale = scaleObj.getLinearScale().nice();

        if(scaleTypes[0] === "time")
            xScale = scaleObj.getTimeScale().nice();

        //if(scaleType === "ordinal")
        //    xScale = scaleObj.getOrdinalScale(); // .nice() method not available

        // set scale range
        xScale.range([dimensions.padding, dimensions.width - dimensions.padding]);

        return xScale;
    }

    getYScale() {
        // get y-values
        const { data, aes } = this.props;
        const yValues       = data.map(item => item[aes[1]]);

        // create xScale
        let scaleObj  = new scaleFinder(yValues);

        return scaleObj;
    }

    getYScaleType() {
        const { scaleTypes, dimensions } = this.props;
        const scaleObj                   = this.getYScale();
        let yScale;

        if(scaleTypes[1] === "linear")
            yScale = scaleObj.getLinearScale().nice();

        if(scaleTypes[1] === "time")
            yScale = scaleObj.getTimeScale().nice();

        //if(scaleType === "ordinal")
        //    yScale = scaleObj.getOrdinalScale(); // .nice() method not available

        // set scale range
        yScale.range([dimensions.height - dimensions.padding, dimensions.padding]);

        return yScale;
    }

    createLine() {
        const { aes } = this.props;
        const xScale  = this.getXScaleType();
        const yScale  = this.getYScaleType();

        // create line for chart
        const chartLine = line()
            .x(d => xScale(d[aes[0]]))
            .y(d => yScale(d[aes[1]]))
            .curve(curveCatmullRom);

        return chartLine;
    }

    appendLine() {
        const { data, color, lineWidth } = this.props;
        const lineToAppend               = this.createLine();

        // append line to plot
        select(this.node)
            .datum(data)
            .attr("fill", "none")
            .attr("stroke", color)
            .attr("stroke-width", lineWidth)
            .attr("d", lineToAppend);
    }

    render() {
        return(
            <g className={ this.props.className }>
                <path
                    ref={ node => this.node = node }
                />
            </g>
        );
    }

    componentDidMount() {
        this.appendLine();
    }

    componentDidUpdate() {
        this.appendLine();
    }
}

export default Line;
