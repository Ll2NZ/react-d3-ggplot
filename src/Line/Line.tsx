import React from "react";
import { useContext } from "react";
import { ChartContext } from "../_context/";
import { useScale } from "../_hooks/";
import { line } from "d3-shape";
import { curveCatmullRom } from "d3-shape";

interface IProps {
  fill?: string;
  width?: string;
}

Line.displayName = "Line";

function Line(props: IProps) {
  const context = useContext(ChartContext);
  const xScale: any = useScale("XAxis");
  const yScale: any = useScale("YAxis");

  const lineToAppend = line()
    .x((d: any) => xScale(d[context.aes[0]]))
    .y((d: any) => yScale(d[context.aes[1]]))
    .curve(curveCatmullRom)(context.data);

  return (
    <g>
      <path
        fill="none"
        stroke={!props.fill ? "crimson" : props.fill}
        strokeWidth={!props.width ? "1" : props.width}
        d={!lineToAppend ? undefined : lineToAppend}
      />
    </g>
  );
}

export default Line;
