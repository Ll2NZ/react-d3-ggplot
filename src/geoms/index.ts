export { default as GEOMS } from "./geoms";
