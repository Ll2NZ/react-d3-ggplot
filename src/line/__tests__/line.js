import React from 'react';
import { render } from '@testing-library/react';
import { cleanup } from '@testing-library/react';
import { GEOMS } from '../../geoms';
import Line from '../line.js';

afterEach(cleanup);

test('Underlying SVG markup should not change when we change our component logic', () => {
  // changing validState will change the markup
  const validState = {
    data: [{ x: 0, y: 0 }, { x: 10, y: 10 }],
    aes: ['x', 'y'],
    dimensions: { width: 500, height: 300, padding: 50 }
  };

  // should render an <svg/> containing a <path/> element
  const { container } = render(
    <GEOMS {...validState}>
      <Line />
    </GEOMS>
  );

  expect(container.firstChild).toMatchSnapshot();
});
