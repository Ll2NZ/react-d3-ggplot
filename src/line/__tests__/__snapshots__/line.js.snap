// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Underlying SVG markup should not change when we change our component logic 1`] = `
<svg
  height="300"
  width="500"
>
  <g>
    <path
      d="M50,250L450,50"
      fill="none"
      stroke="orange"
      stroke-width="2"
    />
  </g>
</svg>
`;
